Basic Docs (#1)
programbo/concentration: Game_Doc.md

# Concentration Game
## Description
A simple concentration card game where the user flips cards to reveal it's icon and must find that cards duplicate by flipping another card.

## Components
* [Board Code](./components/board.tsx) / [Doc](#board)
* [Card Code](./components/card.tsx) / [Doc](#card)

> ## Board

Reviewer: It's usually good to talk to someone who is trying to use your project/components. An example of using the Board component is good:

```tsx
<Board size={[4, 3]} />
```

Letting them know the expected shape of the `size` property is very helpful. At a glance, someone (familiar with Typescript) can look at this Interface* and quickly reason it out:

```ts
interface BoardProps {
  size: [width: number, height: number]
}
```

\* I haven't updated the Board component's Interface the this but it would be a good idea.

The Board component handles the main game logic:
* Creating the game-board from width and height values,
* Builds the list of icons and subsequently the list of cards,
* Shuffles the card positions within the list,
* And returns a grid with cards mapped to each position, each card having it's own logic to control whether it is selected or found.

The Board's width must be: even, more than 0 and no more than 6.
```tsx
(width % 2 === 1 || width > 6)
```

## Card

> The Card component contains the values for:

Reviewer: Drop the Card interface in here so the list below can describe the actual variables.

e.g.
```
* found - (optional) true indicates the card is one of a matched pair and should remain revealed for the rest of the game
```

Reviewer: Also include the CardType definition so the `onClick` callback makes sense.

# Concentration Game
## Description
A simple concentration card game where the user flips cards to reveal it's icon and must find that cards duplicate by flipping another card.

## Components
* [Board Code](./components/board.tsx) / [Doc](#board)
* [Card Code](./components/card.tsx) / [Doc](#card)

## Board

> ### Usage
> ```ts
> <Board size={[4, 3]} />
> ```
> 
> The width property must be even, equal/more than 2 and no more than 6. <br>
> The height propery must be equal or more than 1 and less than or equal to 6.
> <br><br>
> The Board component uses the Boardprops interface referenced below.
> 
> ```ts
>   interface BoardProps {
>     size: [width: number, height: number]
>   }
> ```

Reviewer: ```suggestion
### Usage

The Board component uses the `BoardProps` interface referenced below.

interface BoardProps {
  size: [width: number, height: number]
}

<Board size={[4, 3]} />

* The `width` property must be an **even** integer between 2 and 6.
* The `height` property must be an integer between 2 and 6.
```

Made it a little less wordy and revered the order so the Props interface was introduced BEFORE referring to the array positions as "width" and "height".

_Also, please add the syntax highlighting back to the code snippets - I had to remove them because the Suggestions syntax breaks if you include a triple tick (```) in the suggestion_ 😬

The Board component handles the main game logic:
* Creating the game-board from width and height values,
* Builds the list of icons and subsequently the list of cards,
* Shuffles the card positions within the list,
* And returns a grid with cards mapped to each position, each card having it's own logic to control whether it is selected or found.

## Card
```ts

>   type CardType = {
>   icon: string
>   id: string
>   }

Reviewer: ```suggestion
  type CardType = {
    icon: string
    id: string
  }
```

  interface CardProps {
    data: CardType
    selected?: boolean
    found?: boolean
    onClick: (card: CardType) => void
  }
```
The Card component utilizes the CardProps interface which contains the variables for:

> * data: Contains the icon and card id values

Reviewer: ```suggestion
* data: Contains the `icon` and Card `id` values
```

* selected: (optional) True indicates the card has been clicked.

> * found: (optional) True indicates that the matched pair of cards have been found and flags the card to ignore click events.

Reviewer: ```suggestion
* found: (optional) True indicates that the matched pair of cards have been found.
```

The latter half was more about game logic than the component interface.

> * onClick: Returns the Card icon and id when set.

Reviewer: ```suggestion
* onClick: Callback passes back the clicked Card's `icon` and `id`.
```

Be careful when you use the term "return". The function definition (`(card: CardType) => void`) _passes_ a `CardType` but _returns_ `void` (nothing).
